### ISSUE TITLE
Usage with more than one sequence type

### ISSUE THREAD

**Issue by alex-dixon** [serde-rs/serde#1536]
I'm working on a serde implementation of [EDN](https://github.com/edn-format/edn), which has 3 different types of sequences: vectors, lists, and sets. I'm looking for the best way to serialize and deserialize these using serde. 

I started by forking `serde_json`, so I can successfully deserialize a vector into a `Value::Vector` just like a JSON array using ValueVisitor.visit_seq:

```rust
fn visit_seq<V>(self, mut visitor: V) -> Result<Value, V::Error>
                where
                    V: SeqAccess<'de>,
            {
                let mut vec = Vec::new();

                while let Some(elem) = try!(visitor.next_element()) {
                    vec.push(elem);
                }

                Ok(Value::Vector(vec))
            }
```

I keep thinking I should call `visit_seq` when I detect a list in the input stream, but `visit_seq` doesn't seem to be able to know to return a `Value::Vector` vs a `Value::List`.

I'm running into a similar problem on the serialization end, where I can write out the correct `[]` for a vector through `serialize_seq`, but can't figure out a way to disambiguate from other sequence types like list which requires `()` be written out instead:
```rust
fn serialize_seq(self, len: Option<usize>) -> Result<Self::SerializeSeq> {
        if len == Some(0) {
            try!(self
                .formatter
                .begin_vector(&mut self.writer)
                .map_err(Error::io));
            try!(self
                .formatter
                .end_vector(&mut self.writer)
                .map_err(Error::io));
            Ok(Compound::Map {
                ser: self,
                state: State::Empty,
            })
        } else {
            try!(self
                .formatter
                .begin_vector(&mut self.writer)
                .map_err(Error::io));
            Ok(Compound::Map {
                ser: self,
                state: State::First,
            })
        }
    }
```

I'd really appreciate any help. I've been working on it for a few days. I don't feel much closer to a solution than when I started and can't seem to let it go.

**Comment by alex-dixon** [comment]
Here's a playground trying to deserialize into multiple sequence types:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=28b057779d25b894451e92d48378b7b6

At this point I don't know how to use `serde` to accomplish this without modifying `Visitor`, or supplying my own version of it with something like `visit_set`. Alternatively it seems modifying `visit_seq` to take an argument indicating what type of sequence variant should be returned would permit this also. Either way, most the deserialization path depends on `Visitor`, so I'd need to write a`Deserializer` trait with methods that use the new kind of visitor (`serde::de::Visitor + VisitorExt`) and a `Deserialize` trait for that. 

I've been asking around on Rust Discord and the Rust Forums and haven't received any suggestions on how to be able to do this. I `#serde` IRC briefly as well but didn't get any replies.

**Comment by dtolnay** [comment]
You will need to pick some representation of EDN vectors and lists and sets against [this set of Serde data types](https://serde.rs/data-model.html). The mapping you pick is pretty arbitrary and up to you. For example you could decide that an EDN set would be represented as a Serde **struct** called "edn::Set" with a single field called "elements" which is a Serde **seq**.

Then you need to think about four transformations: Serialize (Rust type to Serde data model), Serializer (Serde data model to EDN), Deserializer (EDN to Serde data model), Deserialize (Serde data model to Rust type). The Serializer and Deserializer steps should be straightforward -- Serializer needs to identify when it is being fed your chosen representation for these EDN types and emit them in the appropriate EDN form, and Deserializer needs to parse EDN and produce your chosen representation to the caller's Visitor when it encounters a list or set. For the Serialize and Deserialize steps you could do it either by having your library define its own Set / List etc types with Serialize and Deserialize impls that produce the chosen representations (the toml crate handles its Datetime type this way so check that out as an example), or alternatively by exposing a `serde(with = "...")` function that would let the user use any existing set / list types probably from the standard library:

```rust
#[derive(Serialize, Deserialize)]
struct Demo {
    #[serde(with = "edn::set")]
    my_set: HashSet<String>,
}
```

**Comment by alex-dixon** [comment]
@dtolnay Thanks! That makes sense.

For others, before I received this advice I ended up adding my own Visitor trait (`trait EDNVisitor<'de> :  serde::de::Visitor`). This allows adding methods specific to the data format (`visit_list`, `visit_set`, `visit_vector`) and leaving the impl for serde's Visitor as-is. It did require copying and renaming `Deserialize` and `Deserializer` so the `serde::de::Visitor` constraint could be changed to the format-specific visitor (`EDNVisitor`). This leverages serde's Visitor for data types that directly map to serde's data model, like string, number, boolean, and unit, and format-specific visitor methods when that mapping is less obvious. 

For sequences, copying `SeqAccess` and changing the trait requirements for methods like `next_element` and `next_element_seed` was necessary in order to require elements of the sequence implement the format-specific Deserialize trait. 

All told the traits I needed to modify fit in about 300LOC.
